go/api/base_client.go

```diff
@@ -244,7 +261,16 @@ func (client *baseClient) executeCommandWithRoute(
 		routeBytesPtr,
 		routeBytesCount,
 	)
+	client.mu.Unlock()
+
 	payload := <-resultChannel
+
+	client.mu.Lock()
+	if client.pending != nil {
+		delete(client.pending, resultChannelPtr)
```

Comment:
R and W operations on that map are protected by the same mutex. Brilliant!

Reply:
Actually, we have only W operations here!

Reply:
R operations only happen on `Close`. But since `Close` can't be concurrent, our only choice is to use a single Mutex.

Reply:
Adding new entries to the map is also a Write operation. Without a mutex a concurrent write may cause a collision and data may be lost.

Reply:
Yes, and Go forbids that with `runtime panic: concurrent map writes or reads`.